keps/sig-testing/20190118-breaking-apart-the-kubernetes-test-tarball.md

```diff
@@ -0,0 +1,221 @@
+Many of the motivations for breaking up that tarball echo into this proposal.
+
+Another motivation comes from the Bazel effort. It's possible to build all
+release artifacts solely using Bazel, and there is progress being made on
+supporting cross-compilation of binary artifacts. There is no support for
+combining artifacts for multiple platforms in one Bazel invocation, however, so
+it is not possible to produce a monolith test tarball using Bazel. Separating
+the test tarball would thus help with the "Bazel parity" effort, though it is
+not the sole motivation.
+
+### Goals
+
+- The Kubernetes test artifacts are broken apart such that users only need to
+  download binaries for platforms they're using.
+- Be largely invisible to most developers: everything should just keep working.
+
+### Non-Goals
+
+- Changing the underlying build system. Both the make/shell-based build
+  system and the Bazel-based build system will be affected, but users can
+  continue to use their existing workflow.
+- Removing cruft from the test tarballs. Likely, there are binaries and portable
+  sources no longer being used anywhere, but we won't prune them with this effort.
+- Changing what is released outside of the test tarball.
+
+## Proposal
+
+Instead of building and distributing a single `kubernetes-test.tar.gz` with all
+portable sources and compiled binaries for all platforms, produce many
+platform-specific tarballs, one for each platform defined in
+[`KUBE_TEST_PLATFORMS`](https://github.com/kubernetes/kubernetes/blob/193f659a1cd454b93cbe1e7b1f13b77c21783461/hack/lib/golang.sh#L150-L160):
+
+- `kubernetes-test-linux-amd64.tar.gz`
+- `kubernetes-test-linux-arm.tar.gz`
+- `kubernetes-test-linux-arm64.tar.gz`
+- `kubernetes-test-linux-s390x.tar.gz`
+- `kubernetes-test-linux-ppc64le.tar.gz`
+- `kubernetes-test-darwin-amd64.tar.gz`
+- `kubernetes-test-windows-amd64.tar.gz`
+
+### Internal structure of the test tarball
+
+At present, the Kubernetes test tarball has several components, all
+rooted under a `kubernetes/` top-level directory.
+
+#### Binary artifacts
+
+The test binary artifacts are currently organized into directories divided by platform:
+
+- `platforms/`
+  - `darwin/`, `linux/`, `windows/`
+    - `amd64/`, `arm/`, `arm64/`, `ppc64le/`, `s390x/`
+
+For comparison, the existing platform-specific tarballs
+(such as `kubernetes-client-linux-amd64.tar.gz`) place all binaries under
+a constant path with no platform information: `kubernetes/client/bin/kubectl`.
+
+Scripts (such as `cluster/get-kube-binaries.sh`) [extract these tarballs
+back into platform-specific directories](https://github.com/kubernetes/kubernetes/blob/193f659a1cd454b93cbe1e7b1f13b77c21783461/cluster/get-kube-binaries.sh#L143-L156)
+to support downloading multiple platforms into a single workspace.
+
+The test tarball should follow the lead of the other platform-specific tarballs
+and place the binaries under `test/bin`. We can then reuse the existing
```

Comment:
Hi @ixdy,

If this path is agreed upon and `e2e.test` will be in `test/bin` inside the tarball, I can go ahead and make some changes to the sk8 environment to accommodate this KEP for when it's implemented.